File: server/webapp/WEB-INF/rails.new/app/presenters/api_v2/config/template_config_representer.rb
##########################################################################
# Copyright 2016 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV2
  module Config
    class TemplateConfigRepresenter < ApiV2::BaseRepresenter
      alias_method :template, :represented

      error_representer

      link :self do |opts|
        opts[:url_builder].apiv2_admin_template_url(template_name: template.name.to_s) unless template.name.blank?
      end

      link :doc do |opts|
        'https://api.go.cd/#template-config'
      end

      link :find do |opts|
        opts[:url_builder].apiv2_admin_template_url(template_name: '__template_name__').gsub(/__template_name__/, ':template_name')
      end


      property :errors, exec_context: :decorator, decorator: ApiV2::Config::ErrorRepresenter, skip_parse: true, skip_render: lambda { |object, options| object.empty? }
      property :name, case_insensitive_string: true

      collection :stages,
                 exec_context: :decorator,
                 decorator:    ApiV2::Config::StageRepresenter,
                 expect_hash:  true,
                 class:        com.thoughtworks.go.config.StageConfig

      def stages
        template.getStages() unless template.getStages().isEmpty
      end

      def stages=(value)
        template.getStages().clear()
        value.each { |stage| template.addStageWithoutValidityAssertion(stage) }
      end
    end
  end
end

File: api/api-template-config-v5/src/main/java/com/thoughtworks/go/apiv5/admin/templateconfig/representers/TemplateConfigRepresenter.java
/*
 * Copyright 2020 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.thoughtworks.go.apiv5.admin.templateconfig.representers;

import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.api.representers.ErrorGetter;
import com.thoughtworks.go.api.representers.JsonReader;
import com.thoughtworks.go.apiv10.admin.shared.representers.stages.StageRepresenter;
import com.thoughtworks.go.config.PipelineTemplateConfig;
import com.thoughtworks.go.spark.Routes;

import java.util.HashMap;

public class TemplateConfigRepresenter {

    public static void toJSON(OutputWriter jsonWriter, PipelineTemplateConfig pipelineTemplateConfig) {
        jsonWriter.addLinks(linksWriter -> linksWriter
                .addLink("self", Routes.PipelineTemplateConfig.name(pipelineTemplateConfig.name().toString()))
                .addAbsoluteLink("doc", Routes.PipelineTemplateConfig.DOC)
                .addLink("find", Routes.PipelineTemplateConfig.find()));
        if (!pipelineTemplateConfig.errors().isEmpty()) {
            jsonWriter.addChild("errors", errorWriter -> {
                HashMap<String, String> errorMapping = new HashMap<>();
                new ErrorGetter(new HashMap<>()).toJSON(errorWriter, pipelineTemplateConfig);
            });
        }

        jsonWriter.add("name", pipelineTemplateConfig.name());
        writeStages(jsonWriter, pipelineTemplateConfig);

    }

    private static void writeStages(OutputWriter jsonWriter, PipelineTemplateConfig pipelineTemplateConfig) {
        if (pipelineTemplateConfig.getStages().isEmpty()) {
            jsonWriter.renderNull("stages");
        } else {
            jsonWriter.addChildList("stages", stagesWriter -> StageRepresenter.toJSONArray(stagesWriter, pipelineTemplateConfig));
        }
    }

    public static PipelineTemplateConfig fromJSON(JsonReader jsonReader) {
        PipelineTemplateConfig pipelineTemplateConfig = new PipelineTemplateConfig();
        jsonReader.readStringIfPresent("name", pipelineTemplateConfig::setName);
        setStages(jsonReader, pipelineTemplateConfig);
        return pipelineTemplateConfig;
    }

    private static void setStages(JsonReader jsonReader, PipelineTemplateConfig pipelineTemplateConfig) {
        pipelineTemplateConfig.getStages().clear();
        jsonReader.readArrayIfPresent("stages", stages -> {
            stages.forEach(stage -> {
                pipelineTemplateConfig.addStageWithoutValidityAssertion(StageRepresenter.fromJSON(new JsonReader(stage.getAsJsonObject())));
            });
        });
    }
}
